File: samples/csharp.net/GeopalJobSample/GeopalClient.cs
using System;
using System.Text;
using System.Security.Cryptography;
using System.Collections.Specialized;
using System.Net;

namespace GeopalJobSample
{
	public class GeopalClient
	{
		private const String DEFAULT_PRIVATEKEY = "";
		private const String DEFAULT_EMPLOYEEID = "";
		private const String GEOPAL_BASE_URL = "https://app.geopalsolutions.com/";

		private String uri = "";
		private String employeeId = DEFAULT_EMPLOYEEID;
		private String privateKey = DEFAULT_PRIVATEKEY;

		public GeopalClient ()
		{
		}

		public GeopalClient (String uri)
		{
			setUri(uri);
		}


		public string post(NameValueCollection pairs) 
		{
			WebClient wb = generateWebClient("POST", this.uri);
			return Encoding.ASCII.GetString(wb.UploadValues(getUrl(), "POST", pairs));
		}

		public string put(NameValueCollection pairs)
		{
			WebClient wb = generateWebClient("PUT", this.uri);
			return Encoding.ASCII.GetString(wb.UploadValues(getUrl(), "PUT", pairs));
		}

		public string get(String urlPairs)
		{
			WebClient wb = generateWebClient("GET", this.uri);
			return wb.DownloadString(getUrl()+"?"+urlPairs);
		}

		public string get ()
		{
			WebClient wb = generateWebClient ("GET", this.uri);
			return wb.DownloadString (getUrl ());
		}

		public void downloadFile(String urlPairs, string fileLocation)
		{
			WebClient wb = generateWebClient("GET", this.uri);
			wb.DownloadFile(getUrl()+"?"+urlPairs, fileLocation);
		}

		public void setUri (String uri)
		{
			this.uri = uri;
		}

		public void setEmployeeId (string employeeId)
		{
			this.employeeId = employeeId;
		}

		public void setPrivateKey (string privatekey)
		{
			this.privateKey = privatekey;
		}

		private String getUrl ()
		{
			return GEOPAL_BASE_URL + this.uri;
		}

		private WebClient generateWebClient(String method, String uri)
		{
			method = method.ToLower();
			DateTime now = DateTime.Now;
			String timestamp = now.ToString("ddd, dd MMM yyyy hh:mm:ss ")+"GMT";
			string signature = GetSignature(method + uri + employeeId + timestamp, privateKey);

			WebClient wb = new WebClient();
			WebHeaderCollection wbHeaders = new WebHeaderCollection();
			wbHeaders.Add("GEOPAL_SIGNATURE", signature);
			wbHeaders.Add("GEOPAL_TIMESTAMP", timestamp);
			wbHeaders.Add("GEOPAL_EMPLOYEEID", employeeId);

			wb.Headers = wbHeaders;

			return wb;

		}

		private string GetSignature(string signtext, string privateKey)
		{
			System.Text.ASCIIEncoding encoding = new System.Text.ASCIIEncoding();
			
			byte[] keybytes = encoding.GetBytes(privateKey);
			byte[] signbytes = encoding.GetBytes(signtext);
			
			HMACSHA256 hmacsha256 = new HMACSHA256(keybytes);
			
			return EncodeTo64(ByteToString(hmacsha256.ComputeHash(signbytes)).ToLower());
		}
		
		private string EncodeTo64(string toEncode)
		{
			byte[] toEncodeAsBytes
				= System.Text.ASCIIEncoding.ASCII.GetBytes(toEncode);
			string returnValue
				= System.Convert.ToBase64String(toEncodeAsBytes);
			return returnValue;
		}
		
		private static string ByteToString(byte[] buff)
		{
			string sbinary = "";
			
			for (int i = 0; i < buff.Length; i++)
			{
				sbinary += buff[i].ToString("X2"); // hex format
			}
			return (sbinary);
		}
	}
}

File: samples/vb.net/VBGeopalJob/GeopalClient.vb
Imports System.Text
Imports System.Security.Cryptography
Imports System.Collections.Specialized
Imports System.Net

Namespace GeopalJobSample
	Public Class GeopalClient
		Private Const DEFAULT_PRIVATEKEY As [String] = ""
		Private Const DEFAULT_EMPLOYEEID As [String] = ""
		Private Const GEOPAL_BASE_URL As [String] = "https://app.geopalsolutions.com/"

		Private uri As [String] = ""
		Private employeeId As [String] = DEFAULT_EMPLOYEEID
		Private privateKey As [String] = DEFAULT_PRIVATEKEY

		Public Sub New()
		End Sub

		Public Sub New(uri As [String])
			setUri(uri)
		End Sub


		Public Function post(pairs As NameValueCollection) As String
			Dim wb As WebClient = generateWebClient("POST", Me.uri)
			Return Encoding.ASCII.GetString(wb.UploadValues(getUrl(), "POST", pairs))
		End Function

		Public Function put(pairs As NameValueCollection) As String
			Dim wb As WebClient = generateWebClient("PUT", Me.uri)
			Return Encoding.ASCII.GetString(wb.UploadValues(getUrl(), "PUT", pairs))
		End Function

		Public Function [get](urlPairs As [String]) As String
			Dim wb As WebClient = generateWebClient("GET", Me.uri)
			Return wb.DownloadString(getUrl() & "?" & urlPairs)
		End Function

		Public Function [get]() As String
			Dim wb As WebClient = generateWebClient("GET", Me.uri)
			Return wb.DownloadString(getUrl())
		End Function

		Public Sub downloadFile(urlPairs As [String], fileLocation As String)
			Dim wb As WebClient = generateWebClient("GET", Me.uri)
			wb.DownloadFile(getUrl() & "?" & urlPairs, fileLocation)
		End Sub

		Public Sub setUri(uri As [String])
			Me.uri = uri
		End Sub

		Public Sub setEmployeeId(employeeId As String)
			Me.employeeId = employeeId
		End Sub

		Public Sub setPrivateKey(privatekey As String)
			Me.privateKey = privatekey
		End Sub

		Private Function getUrl() As [String]
			Return GEOPAL_BASE_URL & Me.uri
		End Function

		Private Function generateWebClient(method As [String], uri As [String]) As WebClient
			method = method.ToLower()
			Dim now As DateTime = DateTime.Now
			Dim timestamp As [String] = now.ToString("ddd, dd MMM yyyy hh:mm:ss ") & "GMT"
			Dim signature As String = GetSignature(method & uri & employeeId & timestamp, privateKey)

			Dim wb As New WebClient()
			Dim wbHeaders As New WebHeaderCollection()
			wbHeaders.Add("GEOPAL_SIGNATURE", signature)
			wbHeaders.Add("GEOPAL_TIMESTAMP", timestamp)
			wbHeaders.Add("GEOPAL_EMPLOYEEID", employeeId)

			wb.Headers = wbHeaders

			Return wb

		End Function

		Private Function GetSignature(signtext As String, privateKey As String) As String
			Dim encoding As New System.Text.ASCIIEncoding()

			Dim keybytes As Byte() = encoding.GetBytes(privateKey)
			Dim signbytes As Byte() = encoding.GetBytes(signtext)

			Dim hmacsha256 As New HMACSHA256(keybytes)

			Return EncodeTo64(ByteToString(hmacsha256.ComputeHash(signbytes)).ToLower())
		End Function

		Private Function EncodeTo64(toEncode As String) As String
			Dim toEncodeAsBytes As Byte() = System.Text.ASCIIEncoding.ASCII.GetBytes(toEncode)
			Dim returnValue As String = System.Convert.ToBase64String(toEncodeAsBytes)
			Return returnValue
		End Function

		Private Shared Function ByteToString(buff As Byte()) As String
			Dim sbinary As String = ""

			For i As Integer = 0 To buff.Length - 1
					' hex format
				sbinary += buff(i).ToString("X2")
			Next
			Return (sbinary)
		End Function
	End Class
End Namespace